### PR TITLE
Implement ravel

### PR DESCRIPTION
**Pull request by ejguan**
Related #38349
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46098 Implement ravel**

Doc:
image

Differential Revision: [D24253213](https://our.internmc.facebook.com/intern/diff/D24253213)

### CODE REVIEW

**Comment by ejguan** [comment]
TODO:
Add tests.

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -1624,6 +1624,11 @@ Tensor flatten(const Tensor& self, DimnameList dims, Dimname out_dim) {
   return native::flatten(self, *dims.begin(), *(dims.end() - 1), out_dim);
 }
 
+Tensor ravel(const Tensor& self) {
+  // Same as flatten() with start_dim as 0 and end_dim as -1.
+  return native::flatten(self, 0, -1);
```
Comment (zou3519) [inline aten/src/ATen/native/TensorShape.cpp:1629]: nit: `self.reshape(-1)` avoids some of the computation that goes on inside `flatten`.

Calling `native::flatten` instead of `at::flatten` inside of here means that we won't actually get autograd support. Using at::flatten (or reshape as pointed out above) fixes that problem. You can test this with something like:
```
x = torch.randn(3, 3, requires_grad=True)
out = x.ravel().sum()
out.backward()  # I think this throws an error with the current implementation
```

File: test/test_torch.py [hunk 2]
```diff
@@ -1887,6 +1887,32 @@ def _test_gather(self, cast, test_bounds=True):
         def test_gather(self):
             self._test_gather(self, lambda t: t)
 
+        def test_ravel(self):
```
Comment (zou3519) [inline test/test_torch.py:1890]: You should add an entry to https://github.com/pytorch/pytorch/blob/59414b359d3a624c63032140e0e1fbc654aa6f5f/torch/testing/_internal/common_methods_invocations.py#L593 to auto-generate an autograd test. You'll be able to see the autograd test by running `pytest test/test_autograd.py -v -k "ravel"` locally (the -k "ravel" tells pytest to filter tests by name)

File: test/test_torch.py [hunk 3]
```diff
@@ -1887,6 +1887,32 @@ def _test_gather(self, cast, test_bounds=True):
...
+            tensors = [torch.randn(5, 5, 5, 5),
+                       torch._empty_affine_quantized([5, 5, 5, 5],
+                                                     scale=2,
+                                                     zero_point=3,
+                                                     dtype=torch.quint8)]
+            for src in tensors:
+                flat = src.ravel()
+                self.assertEqual(flat.shape, torch.Size([625]))
+                self.assertEqual(src.view(-1), flat.view(-1))
+
```
Comment (zou3519) [inline test/test_torch.py:1915]: There are two cases for torch.ravel(tensor):
1) tensor is a contiguous tensor (e.g., tensor.is_contiguous() returns True). Then tensor.ravel() will return a view of the original tensor
2) tensor is a non-contiguous tensor. Then tensor.ravel() will return a copy of the original tensor.

We haven't tested (2) yet, that would be nice to add here.

File: test/test_torch.py [hunk 4]
```diff
@@ -1887,6 +1887,32 @@ def _test_gather(self, cast, test_bounds=True):
...
+            # Test both float tensor and quantized tensor
+            tensors = [torch.randn(5, 5, 5, 5),
+                       torch._empty_affine_quantized([5, 5, 5, 5],
+                                                     scale=2,
+                                                     zero_point=3,
+                                                     dtype=torch.quint8)]
+            for src in tensors:
+                flat = src.ravel()
+                self.assertEqual(flat.shape, torch.Size([625]))
+                self.assertEqual(src.view(-1), flat.view(-1))
```
Comment (zou3519) [inline test/test_torch.py:1914]: self.assertEqual checks equality of values. It doesn't actually check if `flat` is a view of `src` or not, which we do want to check, because that is part of the semantics for ravel()

File: torch/_torch_docs.py [hunk 5]
```diff
@@ -6490,6 +6490,25 @@ def merge_dicts(*dicts):
     tensor([ 1.0000,  1.5000,  2.0000])
 """.format(**factory_common_args))
 
+add_docstr(torch.ravel,
+           r"""
+ravel(input) -> Tensor
+
+Return a contiguous flattened tensor.
```
Comment (zou3519) [inline torch/_torch_docs.py:6497]: It's important to note here that  "A copy is made only if needed."

**Review by zou3519 — COMMENTED** [review]
This generally looks good to me. We should add some more test cases:
- we should add an autograd test to make sure the operator works with autograd
- we should add some more cases to the test_torch test to check the behavior of ravel() on a non-contiguous Tensor

File: test/test_torch.py (hunk 3 above)
Comment (zou3519) [reply test/test_torch.py:1915]: To check if a Tensor is a view of another tensor, one can check `tensor._base`. For example:

```
x = torch.randn(3, 3)
x_view = x.view(-1)
x_copy = x.clone()
assert x_view._base is x
assert x_copy._base is None # because x_copy isn't a view
```

File: aten/src/ATen/native/TensorShape.cpp (hunk 1 above)
Comment (ejguan) [reply aten/src/ATen/native/TensorShape.cpp:1629]: You are right. It doesn't really need to go into `flatten`.
But, t's interesting that it doesn't throw any error and gradient for the tensor is correct when I used `native::flatten`.

Comment (zou3519) [reply aten/src/ATen/native/TensorShape.cpp]: Oh, I see what happened now. native::flatten calls at::reshape, and at::reshape is supported by the autograd.

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -590,6 +590,7 @@ def method_tests():
         ('view', (S,), (S,), '1d', (False,)),
         ('view', (), (dont_convert(()),), 'scalar_to_scalar', (False,)),
         ('view', (), (1,), 'scalar_to_1d', (False,)),
+        ('ravel', (S, S, S), NO_ARGS, '', (True,)),
```
Comment (zou3519) [inline torch/testing/_internal/common_methods_invocations.py:593]: I think the (True,) should be (False,): this boolean represents whether or not this operator is supported by JIT autodiff, which is not the same thing as autograd. Since we didn't make any changes to the JIT in this PR to support `ravel` in JIT autodiff, setting this as (True,) should cause some of the JIT tests to fail.

In general it is OK for an operator to not be supported by JIT autodiff (in fact, most operators aren't supported). JIT autodiff is a symbolic differentiation pass that is useful for the JIT when it tries to fuse things in the backwards graph.

File: test/test_torch.py [hunk 7]
```diff
@@ -1887,6 +1887,41 @@ def _test_gather(self, cast, test_bounds=True):
...
+                flat = src.ravel()
+                self.assertEqual(flat.shape, torch.Size([25]))
+                self.assertEqual(src.view(-1), flat.view(-1))
+                self.assertEqual(flat._base, src)
+
+                # Non-continuous Tensor -> Copy
+                nc_src = src.t()
+                nc_flat = nc_src.ravel()
+                self.assertEqual(nc_flat.shape, torch.Size([25]))
+                self.assertEqual(nc_src.reshape(-1), nc_flat.reshape(-1))
```
Comment (zou3519) [inline test/test_torch.py:1922]: nit: No need to reshape nc_flat again. This should be:
`self.assertEqual(nc_flat, nc-src.reshape(-1)`

File: test/test_torch.py [hunk 8]
```diff
@@ -1887,6 +1887,41 @@ def _test_gather(self, cast, test_bounds=True):
...
+            self.assertEqual(zero_dim_tensor.shape, torch.Size([]))
+            self.assertEqual(flat0.shape, torch.Size([1]))
+            self.assertEqual(one_dim_tensor.shape, torch.Size([1]))
+            self.assertEqual(flat1.shape, torch.Size([1]))
+            self.assertEqual(flat0, one_dim_tensor)
+            self.assertEqual(flat0, flat1)
+            self.assertEqual(flat0.shape, flat1.shape)
+
+            # Test both float tensor and quantized tensor
+            tensors = [torch.randn(5, 5),
```
Comment (zou3519) [inline test/test_torch.py:1906]: nit: We should probably test more shapes. The implementation code for ravel is simple enough that I can believe it works for everything, but there's no harm in testing more

Comment (zou3519) [inline test/test_torch.py:1906]: One edge case we should test is a tensor with a zero sized dimension. For example, the following should be true:
```
x = torch.randn([0, 2, 3]) # tensor of size [0, 2, 3]
y = x.ravel()
assert y.shape = ()
```

File: torch/testing/_internal/common_methods_invocations.py (hunk 6 above)
Comment (ejguan) [reply torch/testing/_internal/common_methods_invocations.py:593]: I see. I though autodiff is similar to autograd. Haha, I may need to also check the JIT.

**Review by zou3519 — APPROVED** [review]
LGTM. Some minor nits on testing.

Also, we should try to test a tensor with a zero size in some dimension. I think the implementation should support it.
